README.md
# mapnik-omnivore
Node module that returns metadata of spatial files.
Version format follows [Semantic Version](http://semver.org/)

[![Build Status](https://travis-ci.org/mapbox/mapnik-omnivore.svg?branch=master)](https://travis-ci.org/mapbox/mapnik-omnivore)
[![Build status](https://ci.appveyor.com/api/projects/status/kr1qems47ex9wp3a?svg=true)](https://ci.appveyor.com/project/Mapbox/mapnik-omnivore)
[![Coverage Status](https://coveralls.io/repos/mapbox/mapnik-omnivore/badge.svg?branch=master)](https://coveralls.io/r/mapbox/mapnik-omnivore?branch=master)

Currently supports the following file types:
- `geojson`
- `topojson`
- `kml`
- `gpx`
- `tif`
- `vrt` : raster files listed in the VRT file must be in the same directory as the VRT file
- `csv` : must be valid geo CSV, and can be in the form of `csv`, `txt`, or `tsv`
- `shp` : In order to set the projection, the `.prj` file must be in the same directory and have the same name as the `.shp` file

## Calculating zoom levels for tiling

For vector datasets:

- calculate `smallestMinZoom` and `smallestMaxZoom`
- starting from zoom level 22 start loop decrementing zoom level by 1 in every iteration
- count how many tiles should be rendered to cover dataset extent fo current zoom level
- calculate metric to asses if current zoom level is good enough
- metric is defined as file size (in bytes) dived by the number of tiles from step 3 and is rough estimate how many data features will be on single tile
- test if metric is below max zoom threshold (if yes then store current zoom level as candidate for max zoom)
- test if metric is below min zoom threshold (if yes then store current zoom level as candidate for min zoom and break zoom level loop)
- start next zoom level loop iteration (decrementing zoom level by 1)
- when zoom level loop will end (by iterating through all zoom levels of by exiting loop)
   - choose max(candidate for max zoom, smallestMaxZoom) as max zoom
   - choose min(candidate for min zoom, smallestMinZoom) as min zoom
   - choose min(min zom, `DEFAULT_SMALLEST_MIN_ZOOM`) as min zoom for file with size under `SMALLEST_MIN_ZOOM_FILE_SIZE`
   - finally set min zoom as min(min zoom, max zoom) and max zoom as max(min_zoom, max_zoom)

## Environment

- `DEFAULT_SMALLEST_MAX_ZOOM` - the smallest max zoom a datasource should be tiled to [`6`]
- `DEFAULT_SMALLEST_MIN_ZOOM` - the smallest min zoom a datasource should be tiled from [`4`]
- `SMALLEST_MAX_ZOOM_FILE_SIZE` - upper file size limit to use max zoom overriding based on gemetry type (max zoom for files above this limit will not be overriden)
- `SMALLEST_MIN_ZOOM_FILE_SIZE` - upper file size limit to use min zoom overriding (min zoom for files above this limit will not be overriden)

## Javascript Usage

```
var mapnikOmnivore = require('@mapbox/mapnik-omnivore'),
    path = require('path');

var file = path.resolve('test/data/zip/world_merc/world_merc.shp');

mapnikOmnivore.digest(file, function(err, metadata){
	if (err) return callback(err);
	else {
		console.log('Metadata returned!');
		console.log(metadata);
	}
});
```

### mapnikOmnivore.digest(filepath, callback)
- filepath `required`
- callback `(err, metadata)`


#### Example of returned `metadata`
```
{ filesize: 428328,   // size of file in bytes
  projection: '+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0.0 +k=1.0 +units=m +nadgrids=@null +wktext +no_defs +over',
  filename: 'world_merc',
  center: [ 0, 12.048603815490733 ],
  extent: [ -180, -59.47306100000001, 180, 83.57026863098147 ],
  json: { 
  	vector_layers: [ { 
      id: 'world_merc',
      description: '',
      minzoom: 0,
      maxzoom: 22,
      fields: { 
        FIPS: 'String',
        ISO2: 'String',
        ISO3: 'String',
        UN: 'Number',
        NAME: 'String',
        AREA: 'Number',
        POP2005: 'Number',
        REGION: 'Number',
        SUBREGION: 'Number',
        LON: 'Number',
        LAT: 'Number' } 
    } ] 
  },
  minzoom: 0,    // calculates the optimal minimum and
  maxzoom: 5,   // maximum zoom levels for the file
  layers: [ 'world_merc' ],
  dstype: 'shape',
  filetype: '.shp' }
```

Raster files will include a `raster` object:
```
{ filesize: 1494,
  projection: '+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=23 +lon_0=-96 +x_0=0 +y_0=0 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs',
  raster: 
   { pixelSize: [ 7.502071930146189, 7.502071930145942 ],
     bandCount: 1,
     bands: [ [Object] ],
     nodata: null,
     origin: [ -1134675.2952829634, 2485710.4658232867 ],
     width: 984,
     height: 804,
     units: { linear: [Object], angular: [Object] } },
  filename: 'sample',
  center: [ -110.32476292309875, 44.56502238336985 ],
  extent: 
   [ -110.3650933429331,
     44.53327824851143,
     -110.28443250326441,
     44.596766518228264 ],
  minzoom: 0,
  maxzoom: 13,
  dstype: 'gdal',
  filetype: '.vrt',
  layers: [ 'sample' ] }

```

## CLI Usage
```
$ npm install --global @mapbox/mapnik-omnivore
$ digest <filepath>
# Prints a JSON string
```

## Generating Mapnik XML

1) Install the https://github.com/mapbox/tilelive-omnivore

2) Use the command line tool it provides, called `mapnik-omnivore` to generate Mapnik XML


## Install
```
npm install @mapbox/mapnik-omnivore
```

## Tests
`npm test`


CHANGELOG.md
# Changelog

## 9.0.0

- Try/catch forward projections in raster extent method
- clamp minZoom for vector data to `process.env.DEFAULT_SMALLEST_MIN_ZOOM` (defaults to 4) for files under `process.env.SMALLEST_MIN_ZOOM_FILE_SIZE` (defaults to 10MB)
- update mapnik to 3.7.0
- update gdal to 0.9.6

## 8.4.1

- Increase raster maxzoom to z25

## 8.4.0

- Upgraded to use mapnik 3.6.0

## 8.3.0

- Shorten the CSV errors when Mapnik cannot find proper column headers [#157](https://github.com/mapbox/mapnik-omnivore/issues/157)
- Introduce an internal framework for determining feature-specific zoom levels, just starting with small point datasets for now [#151](https://github.com/mapbox/mapnik-omnivore/issues/151), [#140](https://github.com/mapbox/mapnik-omnivore/issues/140)
- Update to @mapbox/mapbox-file-sniff and @mapbox/sphericalmercator packages
- Update appveyor to Node v4 and v6, no longer testing on Node v0.10.x

## 8.2.2

- Update to collect all feature's properties from GeoJSON files

## 8.2.1

- Increase raster maxzoom to z23

## 8.2.0

- Move module to @mapbox namespace
- Fixes incorrect ft + mi + km coordinate conversion for raster pixel sizes

## 8.1.0

 - Update to node v6; mapnik 3.5.14; gdal 0.9.3; and mapbox-file-sniff 0.5.2

## 8.0.0

 - Will no longer provide a recommended max zoom level less than 6

## 7.4.0

 - Adjust maxzoom limits for rasters. New fillzoom max: 21, New maxzoom: 22

## 7.3.0

 - Updated to use 3.5.0 mapnik

## 7.1.1

 - Upgraded to node-srs@1.1.0 to pull in node-gdal@0.8.0

## 7.1.0

 - Upgraded to node-gdal@0.8.0 and queue-async@1.1.1

## 7.0.0

 - Adjustments to maxzoom calculations for rasters that result in less oversampling and may yield different ranges than previous versions
 - Fixes a bug in shapefile analysis when mapnik cannot parse the file's projection information

## 6.3.2

 - Now handles more types of GeoJSON parsing errors (both datasource creation and when calling `ds.describe()`) (#128)
 - Upgraded to node-srs@1.x
 - Compatibility with latest node-mapnk@3.4.10 / Mapnik v3.0.9

## 2.2.1

 - Updated node-mapnik to 3.1.1
 - Updated node-gdal to 0.3.1
 - Updated node-srs to 0.4.5


test/utils.test.js
var tape = require('tape');
var utils = require('../lib/utils.js');

tape('[CONVERT TO METERS] Should not convert cell sizes in m', function(assert) {
  var cellSize = [10, 10];
  var unitType = 'm';
  assert.deepEqual(utils.convertToMeters(cellSize, unitType), cellSize, 'Should not convert meter input');
  assert.end();
});

tape('[CONVERT TO METERS] Should accurately convert cell sizes from ft', function(assert) {
  var cellSize = [0.5, 0.5];
  var unitType = 'ft';
  var cellSizeM = utils.convertToMeters(cellSize, unitType);
  var expectedM = [0.1524, 0.1524];
  assert.deepEqual(cellSizeM, expectedM, 'Should convert ft to m accurately');
  assert.end();
});

tape('[CONVERT TO METERS] Should accurately convert cell sizes from miles', function(assert) {
  var cellSize = [500, 500];
  var unitType = 'mi';
  var cellSizeM = utils.convertToMeters(cellSize, unitType);
  var expectedM = [804670, 804670];
  assert.deepEqual(cellSizeM, expectedM, 'Should convert mi to m accurately');
  assert.end();
});

tape('[CONVERT TO METERS] Should accurately convert cell sizes from km', function(assert) {
  var cellSize = [1, 1];
  var unitType = 'km';
  var cellSizeM = utils.convertToMeters(cellSize, unitType);
  var expectedM =  [1000, 1000];
  assert.deepEqual(cellSizeM, expectedM, 'Should convert km to m accurately');
  assert.end();
});

tape('[CONVERT TO METERS] Should accurately convert cell sizes from us-ft', function(assert) {
  var cellSize = [0.5, 0.5];
  var unitType = 'us-ft';
  var cellSizeM = utils.convertToMeters(cellSize, unitType);
  var expectedM = [0.1524, 0.1524];
  assert.deepEqual(cellSizeM, expectedM, 'Should convert us-ft to m accurately');
  assert.end();
});

tape('[CONVERT TO METERS] Should accurately convert cell sizes from us miles', function(assert) {
  var cellSize = [500, 500];
  var unitType = 'us-mi';
  var cellSizeM = utils.convertToMeters(cellSize, unitType);
  var expectedM = [804670, 804670];
  assert.deepEqual(cellSizeM, expectedM, 'Should convert us-mi to m accurately');

  assert.end();
});

tape('[CONVERT TO METERS] Should throw with invalid unit type', function(assert) {
  var cellSize = [500, 500];
  var unitType = 'cubits';
  assert.throws(function() { utils.convertToMeters(cellSize, unitType); });

  assert.end();
});

tape('[CONVERT TO METERS] Should accurately convert cell sizes from decimal degrees', function(assert) {
  var cellSize = [0.0002, 0.0002];
  var unitType = 'decimal degrees';
  var cellSizeM = utils.convertToMeters(cellSize, unitType);
  var expectedM = [22.263888888888893, 22.263888888888893];
  assert.deepLooseEqual(cellSizeM, expectedM, 'Should convert decimal degrees to m accurately');

  assert.end();
});

tape('[SPATIAL RESOLUTIONS] Get spatial resolutions / valid spatial resolutions', function(assert) {
  var spatialResolutions = utils.getSpatialResolutions();
  var expectedResolutions = JSON.parse('[156542.96875,78271.484375,39135.7421875,19567.87109375,9783.935546875,4891.9677734375,2445.98388671875,1222.991943359375,611.4959716796875,305.74798583984375,152.87399291992188,76.43699645996094,38.21849822998047,19.109249114990234,9.554624557495117,4.777312278747559,2.3886561393737793,1.1943280696868896,0.5971640348434448,0.2985820174217224,0.1492910087108612,0.0746455043554306,0.0373227521777153, 0.01866137608885765, 0.009330688044428825 ]');
  assert.deepLooseEqual(spatialResolutions, expectedResolutions);
  var validSpatialResolutions = utils.getValidSpatialResolutions(spatialResolutions, 30.20012, 0);
  var expectedValidResolutions = JSON.parse('[156542.96875,78271.484375,39135.7421875,19567.87109375,9783.935546875,4891.9677734375,2445.98388671875,1222.991943359375,611.4959716796875,305.74798583984375,152.87399291992188,76.43699645996094,38.21849822998047]');
  assert.deepLooseEqual(validSpatialResolutions, expectedValidResolutions);

  assert.end();
});

tape('[SPATIAL RESOLUTIONS] Get spatial resolutions / valid spatial resolutions with weight 0.1 (over)', function(assert) {
  var spatialResolutions = utils.getSpatialResolutions();
  var expectedResolutions = JSON.parse('[156542.96875,78271.484375,39135.7421875,19567.87109375,9783.935546875,4891.9677734375,2445.98388671875,1222.991943359375,611.4959716796875,305.74798583984375,152.87399291992188,76.43699645996094,38.21849822998047,19.109249114990234,9.554624557495117,4.777312278747559,2.3886561393737793,1.1943280696868896,0.5971640348434448,0.2985820174217224,0.1492910087108612,0.0746455043554306,0.0373227521777153, 0.01866137608885765, 0.009330688044428825 ]');
  assert.deepLooseEqual(spatialResolutions, expectedResolutions);

  var validSpatialResolutions = utils.getValidSpatialResolutions(spatialResolutions, 0.1492910087108612, 0.1);
  var expectedValidResolutions = JSON.parse('[156542.96875,78271.484375,39135.7421875,19567.87109375,9783.935546875,4891.9677734375,2445.98388671875,1222.991943359375,611.4959716796875,305.74798583984375,152.87399291992188,76.43699645996094,38.21849822998047,19.109249114990234,9.554624557495117,4.777312278747559,2.3886561393737793,1.1943280696868896,0.5971640348434448,0.2985820174217224]');
  assert.deepLooseEqual(validSpatialResolutions, expectedValidResolutions);

  assert.end();
});

tape('[SPATIAL RESOLUTIONS] Get spatial resolutions / valid spatial resolutions with weight 0.1 (under)', function(assert) {
  var spatialResolutions = utils.getSpatialResolutions();
  var expectedResolutions = JSON.parse('[156542.96875,78271.484375,39135.7421875,19567.87109375,9783.935546875,4891.9677734375,2445.98388671875,1222.991943359375,611.4959716796875,305.74798583984375,152.87399291992188,76.43699645996094,38.21849822998047,19.109249114990234,9.554624557495117,4.777312278747559,2.3886561393737793,1.1943280696868896,0.5971640348434448,0.2985820174217224, 0.1492910087108612, 0.0746455043554306,0.0373227521777153, 0.01866137608885765, 0.009330688044428825 ]');
  assert.deepLooseEqual(spatialResolutions, expectedResolutions);

  var validSpatialResolutions = utils.getValidSpatialResolutions(spatialResolutions, 0.1392910087108612, 0.1);
  var expectedValidResolutions = JSON.parse('[156542.96875,78271.484375,39135.7421875,19567.87109375,9783.935546875,4891.9677734375,2445.98388671875,1222.991943359375,611.4959716796875,305.74798583984375,152.87399291992188,76.43699645996094,38.21849822998047,19.109249114990234,9.554624557495117,4.777312278747559,2.3886561393737793,1.1943280696868896,0.5971640348434448,0.2985820174217224,0.1492910087108612]');
  assert.deepLooseEqual(validSpatialResolutions, expectedValidResolutions);

  assert.end();
});

tape('[SPATIAL RESOLUTIONS] Get spatial resolutions / valid spatial resolutions with weight 0.25', function(assert) {
  var spatialResolutions = utils.getSpatialResolutions();
  var expectedResolutions = JSON.parse('[156542.96875,78271.484375,39135.7421875,19567.87109375,9783.935546875,4891.9677734375,2445.98388671875,1222.991943359375,611.4959716796875,305.74798583984375,152.87399291992188,76.43699645996094,38.21849822998047,19.109249114990234,9.554624557495117,4.777312278747559,2.3886561393737793,1.1943280696868896,0.5971640348434448,0.2985820174217224, 0.1492910087108612,0.0746455043554306,0.0373227521777153, 0.01866137608885765, 0.009330688044428825 ]');
  assert.deepLooseEqual(spatialResolutions, expectedResolutions);

  var validSpatialResolutions = utils.getValidSpatialResolutions(spatialResolutions, 40.20012, 0.25);
  var expectedValidResolutions = JSON.parse('[156542.96875,78271.484375,39135.7421875,19567.87109375,9783.935546875,4891.9677734375,2445.98388671875,1222.991943359375,611.4959716796875,305.74798583984375,152.87399291992188,76.43699645996094]');
  assert.deepLooseEqual(validSpatialResolutions, expectedValidResolutions);

  assert.end();
});

tape('[SPATIAL RESOLUTIONS] Get spatial resolutions / valid spatial resolutions with weight 1', function(assert) {
  var spatialResolutions = utils.getSpatialResolutions();
  var expectedResolutions = JSON.parse('[156542.96875,78271.484375,39135.7421875,19567.87109375,9783.935546875,4891.9677734375,2445.98388671875,1222.991943359375,611.4959716796875,305.74798583984375,152.87399291992188,76.43699645996094,38.21849822998047,19.109249114990234,9.554624557495117,4.777312278747559,2.3886561393737793,1.1943280696868896,0.5971640348434448,0.2985820174217224, 0.1492910087108612, 0.0746455043554306, 0.0373227521777153, 0.01866137608885765, 0.009330688044428825 ]');
  assert.deepLooseEqual(spatialResolutions, expectedResolutions);

  var validSpatialResolutions = utils.getValidSpatialResolutions(spatialResolutions, 39, 1);
  var expectedValidResolutions = JSON.parse('[156542.96875,78271.484375,39135.7421875,19567.87109375,9783.935546875,4891.9677734375,2445.98388671875,1222.991943359375,611.4959716796875,305.74798583984375,152.87399291992188]');
  assert.deepLooseEqual(validSpatialResolutions, expectedValidResolutions);

  assert.end();
});

tape('[DYNAMIC MINZOOM] Get dynamic minZoom / small raster', function(assert) {
  var spatialResolutions = utils.getSpatialResolutions();
  var tileSize = 512;
  var extent = [18.683872388012283, 50.27739459081697, 18.685657559975205, 50.27850451326515];
  var minZoom = utils.getDynamicMinZoom(spatialResolutions, extent, tileSize);
  var expectedMinZoom = 15;
  assert.equals(minZoom, expectedMinZoom);

  assert.end();
});

tape('[DYNAMIC MINZOOM] Get dynamic minZoom / more then size limit raster', function(assert) {
  var spatialResolutions = utils.getSpatialResolutions();
  var tileSize = 512;
  var extent = [17.683872388012283, 48.27739459081697, 18.685657559975205, 50.27850451326515];
  var minZoom = utils.getDynamicMinZoom(spatialResolutions, extent, tileSize);
  var expectedMinZoom = null;
  assert.equals(minZoom, expectedMinZoom);
  assert.end();
});

tape('[DYNAMIC MINZOOM] Get dynamic minZoom / whole world extent', function(assert) {
  var spatialResolutions = utils.getSpatialResolutions();
  var tileSize = 512;
  var extent = [-180, -85.0511287798066, 180, 85.0511287798066];
  var minZoom = utils.getDynamicMinZoom(spatialResolutions, extent, tileSize);
  var expectedMinZoom = null;
  assert.equals(minZoom, expectedMinZoom);
  assert.end();
});


lib/shape.js
var fs = require('fs');
var mapnik = require('mapnik');
var invalid = require('./invalid');
var path = require('path');
var srs = require('srs');
var utils = require('./utils');

module.exports = Shape;

function Shape(filepath) {
  this.filepath = filepath;
  this.basename = path.basename(this.filepath, path.extname(this.filepath));
  this.datasource = new mapnik.Datasource({
    type: 'shape',
    file: filepath,
    layer: this.basename
  });
}

Shape.validFileType = ['shp'];
Shape.prototype.detailsName = 'json';
Shape.prototype.dstype = 'shape';

Shape.prototype.getProjection = function(callback) {
  if (this._projection) return callback(null, this._projection);

  var fileDir = path.dirname(this.filepath);
  var projFile = fileDir + '/' + this.basename + '.prj';
  var _this = this;

  fs.readFile(projFile, function(err, data) {
    if (err) return callback(invalid('Invalid shapefile: missing projection file'));

    // srs and mapnik must be able to parse the projection file
    var result;
    try {
      result = srs.parse(data);
      new mapnik.Projection(result.proj4);
    }
    catch (error) {
      return callback(invalid('Invalid shapefile: invalid projection file'));
    }

    // handle ESRI-specific projections
    if (result.proj4 === undefined) {

      try { result = srs.parse('ESRI::' + data.toString()); }
      catch (error) {
        return callback(
          invalid('Invalid shapefile: invalid projection file')
        );
      }

      if (result.proj4 !== undefined) {
        _this._projection = result.proj4;
        return callback(null, _this._projection);
      } else return callback(invalid('Invalid shapefile: undefined proj4 string'));

    }
    else {
      _this._projection = result.proj4;
      return callback(null, _this._projection);
    }
  });
};

Shape.prototype.getCenter = function(callback) {
  this.getExtent(function(err, extent) {
    if (err) return callback(err);

    return callback(null, [
      0.5 * (extent[0] + extent[2]),
      0.5 * (extent[1] + extent[3])
    ]);
  });
};

Shape.prototype.getExtent = function(callback) {
  if (this._extent) return callback(null, this._extent);

  var _this = this;
  var extent;
  try {
    extent = this.datasource.extent();
  }
  catch (err) {
    return callback(invalid('Invalid shapefile: could not read extent'));
  }

  this.getProjection(function(err, projection) {
    if (err) return callback(err);

    var current = new mapnik.Projection(projection);
    var wgs84 = new mapnik.Projection('+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs');
    var transform;

    if (current !== wgs84) {
      transform = new mapnik.ProjTransform(current, wgs84);
      try {
        extent = transform.forward(extent);
      } catch (err) {
        return callback(invalid('Unable to get extent. Failed to forward projection to WGS84 coordinates'))
      }
    }

    _this._extent = extent;
    return callback(null, _this._extent);
  });
};

Shape.prototype.getDetails = function(callback) {
  // Setup vector layers json
  var actual = this.datasource.describe();
  var fields = actual.fields;
  var json = {
    vector_layers: [
      {
        id: this.basename,
        description: '',
        minzoom: 0,
        maxzoom: 22,
        fields: fields
      }
    ]
  };

  return callback(null, json);
};

Shape.prototype.getLayers = function(callback) {
  return callback(null, [this.basename]);
};

Shape.prototype.getZooms = function(callback) {
  var filepath = this.filepath;
  var ds = this.datasource;

  this.getExtent(function(err, extent) {
    if (err) return callback(err);
    utils.zoomsBySize(filepath, extent, ds, callback);
  });
};


lib/utils.js
var fs = require('fs');
var invalid = require('./invalid');
var SphericalMercator = require('@mapbox/sphericalmercator');
var sm = new SphericalMercator();

var SMALLEST_MAX_ZOOM_FILE_SIZE = process.env.SMALLEST_MAX_ZOOM_FILE_SIZE ? +process.env.SMALLEST_MAX_ZOOM_FILE_SIZE : 1000000; // 1mb
var SMALLEST_MIN_ZOOM_FILE_SIZE = process.env.SMALLEST_MIN_ZOOM_FILE_SIZE ? +process.env.SMALLEST_MIN_ZOOM_FILE_SIZE : 10000000; // 10mb

var DEFAULT_SMALLEST_MAX_ZOOM = process.env.DEFAULT_SMALLEST_MAX_ZOOM ? +process.env.DEFAULT_SMALLEST_MAX_ZOOM : 6;
var DEFAULT_SMALLEST_MIN_ZOOM = process.env.DEFAULT_SMALLEST_MIN_ZOOM ? +process.env.DEFAULT_SMALLEST_MIN_ZOOM : 4;

module.exports.zoomsBySize = function(filepath, extent, datasource, callback) {

  // currently the ogr plugin doesn't pass in a datasource,
  // so we need to redefine the callback
  if (typeof datasource === 'function') callback = datasource;

  var maxSize = 500 * 1024;
  var max = 22;
  var min;

  fs.stat(filepath, function(err, stats) {
    if (err) return callback(err);

    var x;
    var y;
    var z;
    var bounds;
    var tiles;
    var avg;

    // calculate smallest min zoom dynamically
    var smallestMinZoom = getDynamicMinZoom(getSpatialResolutions(), extent, 512, 0.1, Number.MAX_SAFE_INTEGER)
    // set a "smallest max zoom" for different data types, `points` are fixed to maxZoom = 10
    var smallestMaxZoom = stats.size < SMALLEST_MAX_ZOOM_FILE_SIZE ? dataTypeMaxZoom(datasource) : DEFAULT_SMALLEST_MAX_ZOOM;

    for (z = max; z >= 0; z--) {
      // get tile coverage for extent
      bounds = sm.xyz(extent, z, false, 4326);
      // total rows
      x = (bounds.maxX - bounds.minX) + 1;
      // total columns
      y = (bounds.maxY - bounds.minY) + 1;
      // total number of tiles
      tiles = x * y;

      if (tiles <= 0) {
        return callback(invalid('Error calculating min/max zoom: Bounds invalid'));
      }

      if (stats.size <= 0) {
        return callback(invalid('Error calculating min/max zoom: Total bytes less than or equal to zero'));
      }

      // calculate metric value for current zoom
      // metric is rough estimate density of the dataset - how many data features will be on a single tile
      avg = stats.size / tiles;

      if (avg < 1000) {
        // metric is below max zoom threshold, there is no need to tiling zoom levels above current zoom level
        // store current zoom level as candiadate for the max zoom
        max = z;
      }
      if (avg > maxSize) {
        // metric is above min zoom threshold, stop tiling at current zoom level to avoid huge tiles
        // set current zoom level as candidate for the min zoom
        // stop at this zoom level taking into other values that may override zoom level from metric

        // maybe override max zoom candidate
        max = Math.max(max, smallestMaxZoom);
        min = Math.min(smallestMinZoom, z);
        if (stats.size < SMALLEST_MIN_ZOOM_FILE_SIZE) {
          min = Math.min(min, DEFAULT_SMALLEST_MIN_ZOOM);
        }
        min = Math.min(min, max);
        return callback(null, min, max);
      } else if (tiles === 1 || z === 0) {
        // special case for small datasets with one tile or zoom level = 0 only
        max = Math.max(max, smallestMaxZoom);
        min = Math.min(max, smallestMinZoom);
        min = Math.min(min, DEFAULT_SMALLEST_MIN_ZOOM);
        return callback(null, Math.min(min, max), Math.max(min, max));
      }
    }
  });
};

/**
 * Calculates the smallest max zoom a datasource should be tiled to. This is used for small datasources
 * where precision can be lost due to large tile extents. https://github.com/mapbox/mapnik-omnivore/issues/151
 *
 * @param {object} datasource A Node Mapnik datasource object
 * @returns {number} zoom The smallest max zoom to use for the datasource
 */
module.exports.dataTypeMaxZoom = dataTypeMaxZoom;
var dataTypeMaxZoom = function(ds) {

  // try describing the datasource (this catches the ogr hole)
  var info;
  try {
    info = ds.describe();
  } catch (err) {
    return DEFAULT_SMALLEST_MAX_ZOOM;
  }

  // for point features, tile down to z10
  if (info.geometry_type === 'point') {
    return 10;
  }

  // return 6 for all other data types for now
  return DEFAULT_SMALLEST_MAX_ZOOM;
};

module.exports.convertToMeters = function(pixelSize, unit) {
  var circumference = 40075000;
  var conversions = {
    m: function(x) { return x; },

    ft: function(x) { return x * 0.3048; },

    mi: function(x) { return x * 1609.34; },

    km: function(x) { return x * 1000; },

    'us-ft': function(x) { return x * 0.3048; },

    'us-mi': function(x) { return x * 1609.34; },

    'decimal degrees': function(x) { return x / 360 * circumference; }
  };

  if (!conversions[unit]) throw new Error('Invalid unit type, must be one of: [m, ft, mi, km, us-ft, us-mi]');

  var x = conversions[unit](pixelSize[0]);
  var y = conversions[unit](pixelSize[1]);

  return [x, y];
};

module.exports.getUnitType = function(srs) {
  var possibleUnits = ['m', 'ft', 'mi', 'km', 'us-ft', 'us-mi'];
  var i;

  for (i = 0; i < possibleUnits.length; i++) {
    if (srs.indexOf('+units=' + possibleUnits[i]) !== -1) return possibleUnits[i];
  }

  if (srs.indexOf('+units=') === -1 && srs.indexOf('+proj=longlat') !== -1) return 'decimal degrees';

  //Default to meters for now, if nothing matches
  else return 'm';
};

module.exports.getSpatialResolutions = getSpatialResolutions
function getSpatialResolutions() {
  var circumference = 40075000;
  var zoomLevels = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24];

  return zoomLevels.map(function(z) {
    return circumference * Math.cos(0) / Math.pow(2, (z + 8));
  });
};

module.exports.getValidSpatialResolutions = function(spatialResolutions, pixelSize, thresholdWeight) {
  return spatialResolutions.filter(function(res, i) {
    var zBreak = res - spatialResolutions[Math.min(i + 1, spatialResolutions.length - 1)] * thresholdWeight;
    return zBreak > pixelSize;
  });
};

/**
 * Calculates dynamically minZoom for raster sources.
 * Minzoom would be calculated dynamically only for small areas (no more then *sizeLimit*).
 *
 * @param {array} spatialResolutions spatial resolution for every zoom level
 * @param {array} extent raster extent in WGS84
 * @param {number} tileSize size of the tile [px]
 * @returns {number} zoom The smallest min zoom to use for the datasource
 */
function getDynamicMinZoom(spatialResolutions, extent, tileSize, ratio = 0.05, sizeLimit = 50 * 100) {
  // var sizeLimit = 50 * 1000; // 50 km max
  // var ratio = 0.05; // smallest tile area that have to be covered by raster - 5%
  var ex = sm.convert(extent, '900913');
  var w = ex[2] - ex[0];
  var h = ex[3] - ex[1];
  var maxSize = Math.max(w, h);
  var minSize = Math.min(w, h);

  if (minSize >= sizeLimit) {
    return null;
  }

  var zooms = spatialResolutions.filter(function(res, z) {
    var pxPerMeter = spatialResolutions[z];   // pixels per meter
    var size = maxSize / pxPerMeter;          // pixels occupied by raster width or height [m] in current zoom level
    return size >= tileSize * ratio;
  });

  var minZoom = spatialResolutions.length - zooms.length;

  return minZoom;
};

module.exports.getDynamicMinZoom = getDynamicMinZoom